File: src/main/java/com/my/list/domain/Node.java
package com.my.list.domain;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.sql.Timestamp;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class Node implements MainData {

    private Long id;
    private Long user;
    private String type;
    private Timestamp ctime;
    private Timestamp mtime;
    private String title;
    private String excerpt;
    private Boolean part;
    private Boolean collection;
    private String permission;
    private Boolean nsfw;
    private Boolean like;
    private Boolean hide;
    private String source;
    private String description;
    private String comment;

    public static Node fromSingleNode(MainData s) {
        if (s == null) return null;
        if (s instanceof Node) return (Node) s;
        return new Node(s.getId(), s.getUser(), s.getType(), s.getCtime(), s.getMtime(),
            s.getTitle(), s.getExcerpt(), s.getPart(), s.getCollection(), s.getPermission(),
            s.getNsfw(), s.getLike(), s.getHide(), s.getSource(), s.getDescription(), s.getComment());
    }

    public static Node defaultNode() {
        Node node = new Node();
        node.permission = "private";
        node.part = false;
        node.collection = false;
        node.nsfw = false;
        node.like = false;
        node.hide = false;
        return node;
    }

    @Override
    public String toString() {
        return "Node{" +
            "id=" + id +
            ", user=" + user +
            ", type='" + type + '\'' +
            ", ctime=" + ctime +
            ", mtime=" + mtime +
            ", title='" + title + '\'' +
            ", excerpt='" + excerpt + '\'' +
            ", part=" + part +
            ", collection=" + collection +
            ", permission='" + permission + '\'' +
            ", nsfw=" + nsfw +
            ", like=" + like +
            ", hide=" + hide +
            ", source='" + source + '\'' +
            ", description='" + description + '\'' +
            ", comment='" + comment + '\'' +
            '}';
    }
}

File: src/main/java/com/my/list/domain/Node.kt
package com.my.list.domain

import java.sql.Timestamp

class Node(
    override var id: Long? = null,
    override var user: Long? = null,
    override var type: String? = null,
    override var ctime: Timestamp? = Timestamp(System.currentTimeMillis()),
    override var mtime: Timestamp? = Timestamp(System.currentTimeMillis()),
    override var title: String? = null,
    override var excerpt: String? = null,
    override var part: Boolean? = null,
    override var collection: Boolean? = null,
    override var permission: String? = null,
    override var nsfw: Boolean? = null,
    override var like: Boolean? = null,
    override var hide: Boolean? = null,
    override var source: String? = null,
    override var description: String? = null,
    override var comment: String? = null
): MainData {
    override fun toString(): String {
        return "Node[$id,$user,$type,$ctime,$mtime,$title,$excerpt,$part,$collection,$permission,$nsfw,$like,$hide,$source,$description,$comment]"
    }
    companion object { 
        fun fromSingleNode(s: MainData?): Node? {
            if (s == null) return null
            if (s is Node) return s
            return Node(s.id, s.user, s.type, s.ctime, s.mtime, s.title, s.excerpt, s.part, s.collection, s.permission, s.nsfw, s.like, s.hide, s.source, s.description, s.comment)
        }
        fun defaultNode(): Node {
            val node = Node()
            node.permission = "private"
            node.part = false
            node.collection = false
            node.nsfw = false
            node.like = false
            node.hide = false
            return node
        }
    }
}
